src/serac/numerics/functional/tensor.hpp

```diff
@@ -1185,43 +1202,188 @@ SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const t
 
-    swap(b[max_row], b[i]);
-    swap(A[max_row], A[i]);
+    swap(P[max_row], P[i]);
+    swap(U[max_row], U[i]);
+  }
 
-    // zero entries below in this column
+  for (int i = 0; i < n; i++) {
+    // zero entries below in this column in U
+    // and fill in L entries
     for (int j = i + 1; j < n; j++) {
-      double c = -A[j][i] / A[i][i];
-      A[j] += c * A[i];
-      b[j] += c * b[i];
-      A[j][i] = 0;
+      auto c  = U[j][i] / U[i][i];
+      L[j][i] = c;
+      U[j] -= c * U[i];
+      U[j][i] = T{};
     }
   }
 
-  // Solve equation Ax=b for an upper triangular matrix A
-  for (int i = n - 1; i >= 0; i--) {
-    x[i] = b[i] / A[i][i];
-    for (int j = i - 1; j >= 0; j--) {
-      b[j] -= A[j][i] * x[i];
+  return {P, L, U};
+}
+
+/**
+ * @brief Solves a lower triangular system Ly = b
+ *
+ * L must be lower triangular and normalized such that the
+ * diagonal entries are unity. This is not checked in the
+ * function, so failure to obey this will produce
+ * meaningless results.
+ *
+ * @param[in] L A lower triangular matrix
+ * @param[in] b The right hand side
+ * @param[in] P A list of indices to index into b in a permuted fashion.
+ *
+ * @return y the solution vector
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_lower_triangular(const tensor<T, n, n>& L, const tensor<T, n, m...>& b,
+                                                        const tensor<int, n>& P)
+{
+  tensor<T, n, m...> y{};
+  for (int i = 0; i < n; i++) {
+    auto c = b[P[i]];
+    for (int j = 0; j < i; j++) {
+      c -= L[i][j] * y[j];
     }
+    y[i] = c / L[i][i];
   }
+  return y;
+}
 
+/**
+ * @overload
+ * @note For the case when no permutation of the rows is needed.
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_lower_triangular(const tensor<T, n, n>& L, const tensor<T, n, m...>& b)
+{
+  // no permutation provided, so just map each equation to itself
+  // TODO make a convienience function for ranges like this
+  // BT 05/09/2022
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
+
+  return solve_lower_triangular(L, b, P);
+}
+
+/**
+ * @brief Solves an upper triangular system Ux = y
+ *
+ * U must be upper triangular. This is not checked, so
+ * failure to obey this will produce meaningless results.
+ *
+ * @param[in] U An upper triangular matrix
+ * @param[in] y The right hand side
+ * @return x the solution vector
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_upper_triangular(const tensor<T, n, n>& U, const tensor<T, n, m...>& y)
+{
+  tensor<T, n, m...> x{};
+  for (int i = n - 1; i >= 0; i--) {
+    auto c = y[i];
+    for (int j = i + 1; j < n; j++) {
+      c -= U[i][j] * x[j];
+    }
+    x[i] = c / U[i][i];
+  }
   return x;
 }
 
+/**
+ * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
+ * @param[in] A The coefficient matrix A
+ * @param[in] b The righthand side vector b
```

Comment:
```suggestion
 * @param[in] b The right hand side vector b
```

Reply:
I agree with this style, but "righthand" and "lefthand" are used throughout `tensor.hpp`, and I think consistency is more important than the whitespace. I would prefer to leave it as is.